File: frontend/src/api/deduplication.ts
import { apiGet, apiPost } from './client'
import type { DedupStats, PendingReview, DedupDecision } from '@/types/deduplication'
import type { PagedResult } from '@/types/finding'

/** GET /deduplication/stats — get deduplication statistics. */
export function getStats(): Promise<DedupStats> {
  return apiGet<DedupStats>('/deduplication/stats')
}

/** GET /deduplication/pending — list pending duplicate reviews. */
export function listPending(
  page = 1,
  perPage = 25,
): Promise<PagedResult<PendingReview>> {
  const params: Record<string, string> = {
    page: String(page),
    per_page: String(perPage),
  }

  return apiGet<PagedResult<PendingReview>>('/deduplication/pending', params)
}

/** GET /deduplication/history — list deduplication decision history. */
export function listHistory(
  page = 1,
  perPage = 25,
): Promise<PagedResult<DedupDecision>> {
  const params: Record<string, string> = {
    page: String(page),
    per_page: String(perPage),
  }

  return apiGet<PagedResult<DedupDecision>>('/deduplication/history', params)
}

/** POST /deduplication/:relationship_id/confirm — confirm a duplicate relationship. */
export function confirm(relationshipId: string): Promise<void> {
  return apiPost<void>(`/deduplication/${relationshipId}/confirm`, {})
}

/** POST /deduplication/:relationship_id/reject — reject a duplicate relationship. */
export function reject(relationshipId: string): Promise<void> {
  return apiPost<void>(`/deduplication/${relationshipId}/reject`, {})
}

File: backend/src/routes/deduplication.rs
//! Deduplication dashboard API routes.
//!
//! Provides endpoints for viewing duplicate-pair statistics, pending reviews,
//! decision history, and confirming or rejecting duplicate relationships.

use axum::{
    extract::{Path, Query, State},
    Json,
};
use uuid::Uuid;

use crate::errors::{ApiResponse, AppError};
use crate::middleware::auth::CurrentUser;
use crate::middleware::rbac::RequireAnalyst;
use crate::models::pagination::{PagedResult, Pagination};
use crate::services::dedup_dashboard::{self, DedupDecision, DedupStats, PendingReview};
use crate::AppState;

/// GET /api/v1/deduplication/stats -- aggregated dedup statistics.
pub async fn stats(
    State(state): State<AppState>,
    _user: CurrentUser,
) -> Result<Json<ApiResponse<DedupStats>>, AppError> {
    let result = dedup_dashboard::get_stats(&state.db).await?;
    Ok(ApiResponse::success(result))
}

/// GET /api/v1/deduplication/pending -- paginated pending duplicate pairs.
pub async fn pending(
    State(state): State<AppState>,
    _user: CurrentUser,
    Query(pagination): Query<Pagination>,
) -> Result<Json<ApiResponse<PagedResult<PendingReview>>>, AppError> {
    let result = dedup_dashboard::list_pending(&state.db, &pagination).await?;
    Ok(ApiResponse::success(result))
}

/// GET /api/v1/deduplication/history -- paginated decision history.
pub async fn history(
    State(state): State<AppState>,
    _user: CurrentUser,
    Query(pagination): Query<Pagination>,
) -> Result<Json<ApiResponse<PagedResult<DedupDecision>>>, AppError> {
    let result = dedup_dashboard::list_history(&state.db, &pagination).await?;
    Ok(ApiResponse::success(result))
}

/// POST /api/v1/deduplication/{relationship_id}/confirm -- analyst confirms a duplicate.
pub async fn confirm(
    State(state): State<AppState>,
    RequireAnalyst(analyst): RequireAnalyst,
    Path(relationship_id): Path<Uuid>,
) -> Result<Json<ApiResponse<()>>, AppError> {
    dedup_dashboard::confirm(&state.db, relationship_id, analyst.id).await?;
    Ok(ApiResponse::success(()))
}

/// POST /api/v1/deduplication/{relationship_id}/reject -- analyst rejects a duplicate.
pub async fn reject(
    State(state): State<AppState>,
    RequireAnalyst(analyst): RequireAnalyst,
    Path(relationship_id): Path<Uuid>,
) -> Result<Json<ApiResponse<()>>, AppError> {
    dedup_dashboard::reject(&state.db, relationship_id, analyst.id).await?;
    Ok(ApiResponse::success(()))
}
